scraper/ClassScraper.ts

```diff
@@ -0,0 +1,390 @@
+import {
+  reverseDayAndMonth,
+  formatDates,
+  removeHtmlSpecials,
+  makeClassWarning,
+} from './helper'
+import {
+  Term,
+  Class,
+  Status,
+  ClassWarnings,
+  Chunk,
+  Time,
+  Day,
+  ClassTermFinderReference,
+  ClassTermFinderDates,
+  WarningTag,
+} from './interfaces'
+
+interface ClassTermFinderParams {
+  cls: Class
+  reference?: ClassTermFinderReference
+}
+
+/**
+ * @constant { ClassTermFinderReference }: Default reference to follow
+ */
+const defaultReferenceDates: ClassTermFinderReference = [
+  {
+    term: Term.Summer,
+    dates: [{ start: 11, length: 3 }, { start: 12, length: 2 }],
+  },
+  {
+    term: Term.T1,
+    dates: [
+      { start: 1, length: 2 },
+      { start: 1, length: 3 },
+      { start: 1, length: 4 },
+      { start: 2, length: 1 },
+      { start: 2, length: 3 },
+      { start: 3, length: 1 },
+    ],
+  },
+  {
+    term: Term.T2,
+    dates: [
+      { start: 3, length: 2 },
+      { start: 4, length: 2 },
+      { start: 4, length: 3 },
+      { start: 5, length: 1 },
+      { start: 5, length: 3 },
+      { start: 6, length: 1 },
+      { start: 6, length: 3 },
+      { start: 7, length: 1 },
+      { start: 7, length: 2 },
+      { start: 8, length: 1 },
+    ],
+  },
+  {
+    term: Term.T3,
+    dates: [
+      { start: 8, length: 2 },
+      { start: 8, length: 3 },
+      { start: 9, length: 1 },
+      { start: 9, length: 2 },
+      { start: 9, length: 3 },
+      { start: 10, length: 1 },
+      { start: 10, length: 2 },
+    ],
+  },
+  {
+    term: Term.S1,
+    dates: [{ start: 2, length: 4 }],
+  },
+  { term: Term.S2, dates: [{ start: 7, length: 4 }] },
+]
+
+interface ClassTermFinderCheckerParams {
+  start: Date
+  end: Date
+  refDate: ClassTermFinderDates
+}
+
+/**
+ * Classifies a class according to the given reference date. Checks if the class matches the
+ * the given reference dates
+ * @param {Date} start: start date of the class
+ * @param {Date} end: end date of the class
+ * @param {ClassTermFinderDates} refDate: reference dates to match the start and end dates to
+ */
+const classTermFinderChecker = ({
+  start,
+  end,
+  refDate,
+}: ClassTermFinderCheckerParams): Boolean => {
+  return (
+    start.getMonth() + 1 === refDate.start &&
+    end.getMonth() -
+      start.getMonth() +
+      (end.getFullYear() - start.getFullYear()) * 12 ===
+      refDate.length
+  )
+}
+
+/**
+ * Finds the Term for a class. Term is defined in interfaces.ts
+ * @param { Class } cls: Class to find the term for
+ * @param { ClassTermFinderReference } reference: Refernce dates to find the term.
+ * Format of each element: { term: Term, dates: { start: number[], length: number[] } }
+ * start is an array of possible start dates and length is the number of months the term might run for
+ *
+ * @returns { Term }: Term which the class is from
+ */
+const classTermFinder = ({
```

Comment:
Not a very good function name.

```suggestion
const getTermFromClass = ({
```

Reply:
Updated